Add ERC: Regulated Agent Mandate (#1679)
ethereum/ERCs: ERCS/erc-8226.md

---
eip: 8226
title: Regulated Agent Mandate
description: A compliance delegation layer for AI agents operating on tokenized regulated assets.
author: Ludovico Rossi (@ludovicor-eth) <ludovico@brickken.com>, Dario Lo Buglio (@xaler5) <dario@brickken.com>, Thamer Dridi (@thamerdridi) <thamer@brickken.com>, Nabil El Alami Khalifi (@nabil-brickken) <nabil@brickken.com>
discussions-to: https://ethereum-magicians.org/t/erc-8226-regulated-agent-mandate/28208
status: Draft
type: Standards Track
category: ERC
created: 2026-04-12
requires: 165
---

## Abstract

This standard defines a compliance delegation layer for AI agents operating on tokenized regulated assets. It specifies how a verified principal can delegate scoped, time-bounded, and financially capped authority to an on-chain agent, and how regulated token contracts verify mandate validity through a pre-transfer compliance hook before executing transfers.

> RAMS is agnostic to the agent identity system, the token standard, and the token compliance framework. It is designed to work with any trustless agent identity system that maps wallet addresses to agent identifiers (such as [ERC-8004](./eip-8004.md)), any token standard ([ERC-20](./eip-20.md), [ERC-721](./eip-721.md), [ERC-1155](./eip-1155.md)), and any regulated token standard that implements a pre-transfer compliance check (such as [ERC-7943](./eip-7943.md) or [ERC-3643](./eip-3643.md)).

Reviewer: ```suggestion
Regulated Agent Mandate, or RAMS, is agnostic to the agent identity system, the token standard, and the token compliance framework. It is designed to work with any trustless agent identity system that maps wallet addresses to agent identifiers (such as [ERC-8004](./eip-8004.md)), any token standard ([ERC-20](./eip-20.md), [ERC-721](./eip-721.md), [ERC-1155](./eip-1155.md)), and any regulated token standard that implements a pre-transfer compliance check (such as [ERC-7943](./eip-7943.md) or [ERC-3643](./eip-3643.md)).
```

Expand abbreviations the first time they are used.

## Motivation

> The market for tokenized real-world assets is entering a phase of institutional adoption. Platforms operating under MiCA, VARA, and equivalent regulatory frameworks are beginning to support programmable, agent-driven portfolio management on regulated instruments. AI agents that can autonomously execute securities transactions are no longer theoretical — they are being built now, without a standard that makes their operation legally defensible.

Reviewer: > [...] MiCA, VARA [...]

Same comment. You'll need to expand these. It's also possible that these will qualify as "external links". You might want to just stay generic and say:

```suggestion
The market for tokenized real-world assets is entering a phase of institutional adoption. Platforms operating under regulatory frameworks are beginning to support programmable, agent-driven portfolio management on regulated instruments. AI agents that can autonomously execute securities transactions are no longer theoretical — they are being built now, without a standard that makes their operation legally defensible.
```

An agent purchasing a tokenized fund unit on behalf of an investor must satisfy three conditions that no existing standard addresses jointly:

**(a)** The principal on whose behalf the agent acts must be a verified, KYC-cleared legal identity, not merely an Ethereum address.

**(b)** The mandate granted to the agent must be legally traceable, time-bounded, and financially capped, analogous to a power of attorney in traditional finance.

> **(c)** The asset contract must verify mandate validity atomically at the point of transfer, without relying on off-chain coordination.

Reviewer: Just use a regular markdown list:

```suggestion
1. The principal on whose behalf the agent acts must be a verified, KYC-cleared legal identity, not merely an Ethereum address.
1. The mandate granted to the agent must be legally traceable, time-bounded, and financially capped, analogous to a power of attorney in traditional finance.
1. The asset contract must verify mandate validity atomically at the point of transfer, without relying on off-chain coordination.
```

Compliance frameworks such as [ERC-7943](./eip-7943.md) and investor eligibility standards such as [ERC-3643](./eip-3643.md) govern who may hold or transact a regulated token, but neither defines an agent delegation model. Trustless agent identity standards such as [ERC-8004](./eip-8004.md) provide agent discovery and trust signals but no mandate framework. RAMS defines the delegation interface, the compliance provider model, and the integration pattern with regulated token contracts.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHOULD", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174. All implementations MUST implement [ERC-165](./eip-165.md).

> RAMS defines two interfaces deployed as separate contracts:

Reviewer: Is "separate contracts" a requirement? If so, you should use an UPPERCASE keyword to indicate it.

    /// @notice Revokes a principal's eligibility for a given scope.
    /// @param principal The on-chain address of the principal.
    /// @param scopeHash The keccak256 hash of the off-chain scope document.
    /// @param reason The reason for revocation.
    function revokePrincipal(address principal, bytes32 scopeHash, ReasonCode reason) external;

    /// @notice Returns eligibility of a principal for a given scope.
    /// @param principal The on-chain address of the principal.
    /// @param identityRef An off-chain identity reference (e.g., keccak256 of a DID or attestation ID).
    /// @param scopeHash The keccak256 hash of the off-chain scope document.
    /// @return eligible True if the principal is compliant for this scope.
    /// @return reason Reason code. MUST be COMPLIANT when eligible is true.
    /// @return expiresAt Unix timestamp after which this result MUST be re-checked. 0 means no expiry.
    function checkPrincipal(address principal, bytes32 identityRef, bytes32 scopeHash)
        external view returns (bool eligible, ReasonCode reason, uint48 expiresAt);
}
```

> An `IComplianceProvider` implementation MAY delegate identity verification to on-chain identity standards, EAS attestations, or any other identity backend. The interface is agnostic to the source.

Reviewer: Expand EAS please.

### `IAgentMandate`

`IAgentMandate` is implemented by the RAMS registry, a single contract deployed by a registry operator (e.g., a platform or a regulated entity acting as operator). Principals interact with this contract to grant, extend, and revoke mandates. Regulated token contracts interact with this contract inside their pre-transfer compliance hook to verify mandate validity and record executions.

> Each `agentId` has at most one active mandate at any given time. This constraint is by design: in regulated markets, each agent-principal relationship requires segregated accounts and independent audit trails. An operator serving multiple principals deploys one agent wallet per principal, each with its own `agentId` and mandate. This mirrors the account segregation requirements of MiCA (Article 70), MiFID II (Article 16), and VARA.

Reviewer: > This mirrors the account segregation requirements of MiCA (Article 70), MiFID II (Article 16), and VARA.

This is _definitely_ an external link. We generally don't permit those, with certain exceptions listed in EIP-1.

The easiest option here is to remove the specific reference and summarize the content here. If that is insufficient (and MiCA/VARA don't already fall under one of the exceptions), you can follow the procedure defined in EIP-5757 to apply to have a new source allowed.

> Value limits in `MandateScopeParams` are denominated in the base unit of the token at `assetAddress`. If `assetAddress` is `address(0)` (asset-class mandate), limits are denominated in the smallest unit of the currency declared in the off-chain scope document. `uint128` accommodates any practical financial instrument value with 18-decimal precision (up to approximately 3.4 × 10^20 whole tokens). A value of `type(uint128).max` signals "no limit" for both `maxTransactionValue` and `maxCumulativeValue`. Implementations MUST revert if an incoming amount exceeds `type(uint128).max`.

Reviewer: > [...] the off-chain scope document

This isn't previously discussed (or I missed it!), so you could maybe change it to "an off-chain..." if it's a generally understood thing, or define it earlier in the document if it's specific to this proposal.

> The `EnforcerTier` enum distinguishes platform-initiated from regulator-initiated enforcement. This distinction is legally material: in a regulatory dispute, the audit trail must identify both the enforcer address and the authority tier. Implementations MUST use an access control mechanism (e.g., OpenZeppelin `AccessControl`) that enforces this distinction. Access control for `freezeAgent` and `unfreezeAgent` MUST be enforced by the implementation. Global freeze (`bytes32(0)`) MUST be restricted to `REGULATORY` tier enforcers. Jurisdiction-scoped freeze MAY be executed by either tier. The admin role governing enforcer permissions MUST NOT be the same address as any enforcer.

Reviewer: Don't reference OZ directly.

An approved operator MAY call `revokeMandate` and `extendMandate` on behalf of the principal.

`recordExecution` MUST only be callable by the token at `onChainScope.assetAddress` (asset-specific mandates), or by a contract registered in the RAMS token registry, or by an address with enforcer privileges (asset-class mandates). Arbitrary callers MUST be rejected. `recordExecution` MUST revert if the amount exceeds `maxTransactionValue` (when not set to `type(uint128).max`), or if `cumulativeUsed + amount` exceeds `maxCumulativeValue` (when not set to `type(uint128).max`). `cumulativeUsed` MUST NOT reset on `extendMandate`. A cap reset requires explicit revocation and re-issuance.

`grantMandate` MUST revert if `signature` is empty and `msg.sender != principal`. If `signature` is non-empty, implementations MUST verify the signature against `principal`. `grantMandate` MUST revert if `identityRef` is non-zero and the designated `complianceProvider.checkPrincipal` returns `eligible == false`. `grantMandate` MUST revert if `agentId` already has an active mandate. `extendMandate` MUST revert if `newValidUntil` is less than or equal to the current `validUntil`.

> `isActive` MUST return true if and only if all of the following hold: (a) the mandate exists; (b) `validFrom <= block.timestamp <= validUntil`; (c) the mandate is not revoked; (d) `isFrozen(agentId, mandate.onChainScope.jurisdictionHash)` is false; (e) `isFrozen(agentId, bytes32(0))` is false; (f) `complianceProvider` is `address(0)` or `complianceProvider.checkPrincipal` returns `eligible == true`; (g) `maxCumulativeValue` is `type(uint128).max` or `cumulativeUsed < maxCumulativeValue`. `isActiveForAmount` MUST return true if and only if `isActive` returns true and additionally: (h) `maxTransactionValue` is `type(uint128).max` or `uint128(amount) <= maxTransactionValue`; (i) `maxCumulativeValue` is `type(uint128).max` or `cumulativeUsed + uint128(amount) <= maxCumulativeValue`.

Reviewer: This might be more readable as a pseudocode block, but that's entirely up to you.

    return true;
}
```

```solidity
// Pseudocode: transfer recording RAMS execution.

function transfer(address to, uint256 amount) public returns (bool) {
    require(canTransfer(msg.sender, to, amount), ERC7943CannotTransfer(msg.sender, to, amount));

    uint256 agentId = agentRegistry.getAgentByWallet(msg.sender);
    if (agentId != 0) {
        address principal = ramsRegistry.getActivePrincipal(agentId);
        ramsRegistry.recordExecution(agentId, principal, amount);
        emit AgentOperationDetected(agentId, principal, amount);
    }

    return super.transfer(to, amount);
}
```

RAMS-aware tokens SHOULD emit the following event for every agent-initiated transfer:

```solidity
event AgentOperationDetected(
    uint256 indexed agentId,
    address indexed principal,
    uint256 amount
);
```

The compliance responsibilities across the three layers are as follows:

| Layer | Responsibility | Standard |
|---|---|---|
| Token compliance | Investor eligibility on this specific asset | Token compliance framework (e.g., [ERC-7943](./eip-7943.md), [ERC-3643](./eip-3643.md)) |
| Mandate compliance | Agent authority from this principal for this scope | This ERC |

> | Agent identity | Agent exists and is registered | Agent registry (e.g., [ERC-8004](./eip-8004.md)) |

Reviewer: This seems like it belongs either in Motivation or possibly Rationale. Likely Motivation since it's talking about the whole proposal and not a particular decision within it.

### Scope Document

> The off-chain JSON referenced by `scopeHash` MUST be stored on IPFS or equivalent content-addressed storage. The `notes` field SHOULD describe in human-readable language: (a) the actions delegated, (b) operational limitations not captured by structured fields, and (c) the regulatory context.

Reviewer: Ah, okay. Makes more sense now. You should probably introduce this concept earlier just so the earlier reference makes sense. You don't need to move the whole section, but maybe adding an overview or something would help?

Jurisdiction-scoped freeze is supported because a regulatory action in one jurisdiction must not prevent an agent from operating in others. Global freeze (`bytes32(0)`) is reserved for `REGULATORY` enforcers only, reflecting the exceptional nature of a full regulatory halt.

Operator permissions are explicitly scoped so that delegation of authority remains auditable. An operator can revoke or extend a mandate (defensive actions) but cannot grant new mandates (offensive actions that create new legal obligations). This asymmetry reflects the fiduciary principle that a delegate should be able to limit or terminate authority but not expand it without the principal's direct authorization.

RAMS defines its own `IComplianceProvider` interface that is agnostic to the identity verification backend. An implementation MAY use an [ERC-3643](./eip-3643.md)-compatible registry, an [ERC-7943](./eip-7943.md) compliance module, or any other on-chain identity source as the data backend for principal eligibility checks. There is no dependency on any specific token compliance or identity standard.

## Backwards Compatibility

> RAMS introduces no changes to any existing standard. Regulated tokens that are not RAMS-aware continue to function normally. Non-agent transfers are unaffected. Agent-initiated transfers on tokens that do not consult the RAMS registry pass or fail based on the token's existing investor eligibility logic, independent of any mandate. RAMS-aware tokens extend the internal logic of their pre-transfer compliance hook to detect agents and verify mandates; the external interface of the underlying token standard is unchanged. RAMS is compatible with any regulated token framework that implements a pre-transfer compliance check, including [ERC-7943](./eip-7943.md), [ERC-3643](./eip-3643.md), and custom implementations. RAMS is compatible with any agent identity system that provides a wallet-to-agent mapping, including [ERC-8004](./eip-8004.md).

Reviewer: I think you can trim this to just the first sentence :P

## Reference Implementation

> A reference implementation is under development.

Reviewer: ```suggestion
<!-- TODO: reference implementation -->
```

## Security Considerations

> `identityRef` is a reference, not proof. The compliance provider MUST verify the referenced identity at `grantMandate` time and MUST return `eligible == false` if the attestation is invalid or revoked. Callers MUST NOT treat `identityRef != bytes32(0)` as proof of eligibility without compliance provider verification. If an agent wallet address is also a standard investor address, the pre-transfer agent-detection logic could misidentify it. Agent identity registries MUST require proof of key control for agent wallet registration and SHOULD emit a distinct event when an address is registered as an agent wallet.

Reviewer: Requirements (anything with an UPPERCASE keyword) should be kept to the Specification section, while their security implications should be discussed here. The idea is that an implementer should be able to make a complete (and secure) implementation using only that section.

---
eip: 8226
title: Regulated Agent Mandate
description: A compliance delegation layer for AI agents operating on tokenized regulated assets.
author: Ludovico Rossi (@ludovicor-eth) <ludovico@brickken.com>, Dario Lo Buglio (@xaler5) <dario@brickken.com>, Thamer Dridi (@thamerdridi) <thamer@brickken.com>, Nabil El Alami Khalifi (@nabil-brickken) <nabil@brickken.com>
discussions-to: https://ethereum-magicians.org/t/erc-8226-regulated-agent-mandate/28208
status: Draft
type: Standards Track
category: ERC
created: 2026-04-12
requires: 165
---

## Abstract

This standard defines a compliance delegation layer for AI agents operating on tokenized regulated assets. It specifies how a verified principal can delegate scoped, time-bounded, and financially capped authority to an on-chain agent, and how regulated token contracts verify mandate validity through a pre-transfer compliance hook before executing transfers.

> Regulated Agent Mandate, or RAMS, is agnostic to the agent identity system, the token standard, and the token compliance framework. It is designed to work with any trustless agent identity system that maps wallet addresses to agent identifiers (such as [ERC-8004](./eip-8004.md)), any token standard ([ERC-20](./eip-20.md), [ERC-721](./eip-721.md), [ERC-1155](./eip-1155.md)), and any regulated token standard that implements a pre-transfer compliance check (such as [ERC-7943](./eip-7943.md) or [ERC-3643](./eip-3643.md)).

Reviewer 2: ```suggestion
Regulated Agent Mandate Standard, or RAMS, is agnostic to the agent identity system, the token standard, and the token compliance framework. It is designed to work with any trustless agent identity system that maps wallet addresses to agent identifiers (such as [ERC-8004](./eip-8004.md)), any token standard ([ERC-20](./eip-20.md), [ERC-721](./eip-721.md), [ERC-1155](./eip-1155.md)), and any regulated token standard that implements a pre-transfer compliance check (such as [ERC-7943](./eip-7943.md) or [ERC-3643](./eip-3643.md)).
```

1. The asset contract must verify mandate validity atomically at the point of transfer, without relying on off-chain coordination.

Compliance frameworks such as [ERC-7943](./eip-7943.md) and investor eligibility standards such as [ERC-3643](./eip-3643.md) govern who may hold or transact a regulated token, but neither defines an agent delegation model. Trustless agent identity standards such as [ERC-8004](./eip-8004.md) provide agent discovery and trust signals but no mandate framework. RAMS defines the delegation interface, the compliance provider model, and the integration pattern with regulated token contracts.

The compliance responsibilities across the three layers are as follows:

| Layer | Responsibility | Standard |
|---|---|---|
| Token compliance | Investor eligibility on this specific asset | Token compliance framework (e.g., [ERC-7943](./eip-7943.md), [ERC-3643](./eip-3643.md)) |
| Mandate compliance | Agent authority from this principal for this scope | This ERC |
| Agent identity | Agent exists and is registered | Agent registry (e.g., [ERC-8004](./eip-8004.md)) |

## Specification

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHOULD", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174. All implementations MUST implement [ERC-165](./eip-165.md).

Reviewer 2: ```suggestion
The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHOULD", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119) and [RFC 2119](https://tools.ietf.org/html/rfc8174). All implementations MUST implement [ERC-165](./eip-165.md).
```

## Backwards Compatibility

RAMS introduces no changes to any existing standard.

## Reference Implementation

<!-- TODO: reference implementation -->

## Security Considerations

`identityRef` is a reference, not proof. Eligibility comes from `complianceProvider.checkPrincipal`, not from a non-zero `identityRef` value. If an agent wallet address is also a standard investor address, the pre-transfer agent-detection logic could misidentify it. Agent identity registries should require proof of key control for agent wallet registration and emit a distinct event when an address is registered as an agent wallet.

If `recordExecution` were callable by arbitrary addresses, an attacker could advance `cumulativeUsed` to exhaust the cap and deny service to the legitimate agent. The caller restriction specified in the Specification closes this attack surface. Callers can use `isActiveForAmount` for pre-transaction checks or rely on `recordExecution`'s revert behavior for atomic enforcement.

> A compromised compliance provider can declare non-compliant principals as eligible. Principals should only use compliance providers with audited, time-locked upgrade mechanisms. The `REGULATORY` enforcer tier operates independently of compliance provider state, so a captured provider cannot prevent a regulator from halting an agent. If a compliance provider contract becomes non-responsive, all mandates referencing that provider become inoperative because `isActive` calls `complianceProvider.checkPrincipal`, which will revert. This constitutes a systemic denial-of-service risk. Principals should select compliance providers with documented uptime SLAs and audited fallback mechanisms. Implementations may define a grace period after which a mandate with an unresponsive provider is auto-revoked rather than permanently blocked, provided the revocation is logged for audit.

Reviewer 2: Can you define the SLA abbreviation?
